v1/contracts/Registry.sol

```diff
@@ -0,0 +1,127 @@
+        Pointer storage pointer = derivativePointers[derivativeAddress];
+
+        // Ensure that the pointer was not valid in the past (derivatives cannot be re-registered or double
+        // registered).
+        require(pointer.valid == PointerValidity.Invalid);
+        pointer.valid = PointerValidity.Valid;
+
+        registeredDerivatives.push(RegisteredDerivative(derivativeAddress, msg.sender));
+
+        // No length check necessary because we should never hit (2^127 - 1) derivatives.
+        pointer.index = uint128(registeredDerivatives.length.sub(1));
+
+        // Set up PartiesMap for this derivative.
+        PartiesMap storage partiesMap = derivativesToParties[derivativeAddress];
+        for (uint i = 0; i < parties.length; i = i.add(1)) {
+            partiesMap.parties[parties[i]] = true;
+        }
+
+        address[] memory partiesForEvent = parties;
+        emit RegisterDerivative(derivativeAddress, partiesForEvent);
+    }
+
+    function isDerivativeRegistered(address derivative) external view returns (bool isRegistered) {
+        return derivativePointers[derivative].valid == PointerValidity.Valid;
+    }
+
+    function getRegisteredDerivatives(address party) external view returns (RegisteredDerivative[] memory derivatives) {
+        // This is not ideal - we must statically allocate memory arrays. To be safe, we make a temporary array as long
+        // as registeredDerivatives. We populate it with any derivatives that involve the provided party. Then, we copy
+        // the array over to the return array, which is allocated using the correct size. Note: this is done by double
+        // copying each value rather than storing some referential info (like indices) in memory to reduce the number
+        // of storage reads. This is because storage reads are far more expensive than extra memory space (~100:1).
+        RegisteredDerivative[] memory tmpDerivativeArray = new RegisteredDerivative[](registeredDerivatives.length);
+        uint outputIndex = 0;
+        for (uint i = 0; i < registeredDerivatives.length; i = i.add(1)) {
+            RegisteredDerivative storage derivative = registeredDerivatives[i];
+            if (derivativesToParties[derivative.derivativeAddress].parties[party]) {
+                // Copy selected derivative to the temporary array.
+                tmpDerivativeArray[outputIndex] = derivative;
+                outputIndex = outputIndex.add(1);
+            }
+        }
+
+        // Copy the temp array to the return array that is set to the correct size.
+        derivatives = new RegisteredDerivative[](outputIndex);
+        for (uint j = 0; j < outputIndex; j = j.add(1)) {
+            derivatives[j] = tmpDerivativeArray[j];
+        }
+    }
+
+    function getAllRegisteredDerivatives() external view returns (RegisteredDerivative[] memory derivatives) {
+        return registeredDerivatives;
+    }
+
+    /*
+     * @notice Do not call this function externally.
```

Comment:
Can we note here that if it is called a second time, the MultiRole checks will cause it to revert?

Reply:
Done.